toolchain/check/eval.cpp

```diff
@@ -768,15 +762,115 @@ static auto ReplaceTypeWithConstantValue(EvalContext& eval_context,
   return IsConstant(*phase);
 }
 
+static auto ResolveSpecificDeclForInst(EvalContext& eval_context,
+                                       const SemIR::Inst& inst) -> void;
+
+// Resolve the specific declaration in the given instruction field. There is an
+// overload for any field type that has a GetConstantValue() overload which
+// canonicalizes a specific as part of forming its constant value.
+static auto ResolveSpecificDeclForField(EvalContext& eval_context,
+                                        SemIR::SpecificId specific_id) -> void {
+  if (!specific_id.has_value()) {
+    return;
+  }
+
+  const auto& specific = eval_context.specifics().Get(specific_id);
+  const auto& generic = eval_context.generics().Get(specific.generic_id);
+  if (!generic.decl_block_id.has_value()) {
+    // Impl witness table construction happens before its own generic is
+    // finished, in order to make the table's instructions dependent
+    // instructions of the Impl's generic. So we can not resolve the specific
+    // declaration for such instructions at the time they are created, as the
+    // generic's declaration eval block is not yet constructed (with those
+    // dependent instructions in it).
+    //
+    // TODO: Find a way to avoid this early out by making the evaluation of
+    // instructions in the witness table that require a specific happen after
+    // the generic is complete. Or find a way to limit this just to the impl
+    // witness table and ensure the declarations will be resolved.
+    return;
+  }
+
+  ResolveSpecificDeclaration(eval_context.context(),
```

Comment:
Very minor, and only kind of part of this PR because this is pre-existing, but might want to consistently abbreviate `Decl` (rather than `Declaration`, since we've generally agreed on that abbreviation)

i.e.:

```suggestion
  ResolveSpecificDecl(eval_context.context(),
```

Reply:
Done